### PR TITLE
Feature: Redirect Back to Lesson Page After Logging In

### PR DESCRIPTION
**Pull request by OmarMoataz**
#### Because:
https://github.com/TheOdinProject/theodinproject/issues/1954

#### This commit
Made browser redirect to lesson or specific page the user was on in case they go to sign in from that specific page.

#### Checklist
 - [x] You have read [The Odin Projects Contributing Guide](https://github.com/TheOdinProject/theodinproject/wiki/Contributing-Guide)?
 - [x] You have verified the tests and linters all pass against your changes?
 - [x] You have included automated tests for your changes where applicable?

### CODE REVIEW

**Review by KevinMulhern — CHANGES_REQUESTED** [review]
Hi @OmarMoataz, thanks for all your work on this.

I've just tested it in the review app attached to this PR. It looks like it will redirect to all pages instead of just lesson pages? we only want it to redirect if the last page the user was on is a lesson pages please mate.

**Comment by OmarMoataz** [comment]
> Hi @OmarMoataz, thanks for all your work on this.
> 
> I've just tested it in the review app attached to this PR. It looks like it will redirect to all pages instead of just lesson pages? we only want it to redirect if the last page the user was on is a lesson pages please mate.

Yeah, will work on that. I just thought it was a nice to have feature for all pages.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -39,6 +45,19 @@ def not_found_error
...
+  def store_user_location!
+    store_location_for(:user, request.fullpath)
+  end
+
+  # Its important that the location is NOT stored if:
+  # - The request method is not GET (non idempotent)
+  # - The request is handled by a Devise controller such as Devise::SessionsController as that could cause an
+  #    infinite redirect loop.
+  # - The request is an Ajax request as this can lead to very unexpected behaviour.
+  def storable_location?
```
Comment (KevinMulhern) [inline app/controllers/application_controller.rb:57]: I wonder if we can add the check for the lesson page here? and maybe use the controller name and action of the request. Just incase `lessons` is in the full path for an unrelated page.

We should be able to do something like:
```
def storable_location?
    request.get? && is_navigational_format? && !devise_controller? && !request.xhr? && lesson_page?
 end

def lesson_page?
    controller_path == "lessons" && action_name == "show"
 end
```

Then the `after_sign_in_path_for` method could be simplified to:
```
def after_sign_in_path_for(resource)
    stored_location_for(resource) || dashboard_path
end
```

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -39,6 +45,19 @@ def not_found_error
 
   private
 
+  def store_user_location!
+    store_location_for(:user, request.fullpath)
+  end
+
+  # Its important that the location is NOT stored if:
```
Comment (KevinMulhern) [inline app/controllers/application_controller.rb:52]: Can we remove these comments please?

**Review by KevinMulhern — CHANGES_REQUESTED** [review]
This is looking ace @OmarMoataz, let me know if you have any questions about the feedback. Happy to discuss.

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (OmarMoataz) [reply app/controllers/application_controller.rb:52]: Sure.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (OmarMoataz) [reply app/controllers/application_controller.rb:57]: Yeah, that's a better way to do it for sure.

**Review by KevinMulhern — APPROVED** [review]
Great work @OmarMoataz 💪 this has been something thats frustrated me when using the site for such a long time.

**Pull request merged**